File: typescript/packages/mechanisms/evm/src/batch-settlement/claims.ts
import type { Network } from "@x402/core/types";
import type { BatchSettlementVoucherClaim } from "./types";
import { computeChannelId } from "./utils";
import type { Channel, ChannelStorage } from "./storage/channel";

export type SelectClaimableVouchersOptions = {
  now?: number;
  idleSecs?: number;
};

/**
 * Collects vouchers that are eligible for onchain claiming.
 *
 * A voucher is claimable when its `chargedCumulativeAmount` exceeds what has
 * already been claimed onchain. An optional idle filter skips sessions that
 * received a request within the last `idleSecs` seconds. Input order is preserved.
 *
 * @param channels - Channel records to inspect.
 * @param opts - Optional wall-clock and idle filter.
 * @returns Claimable voucher payloads.
 */
export function selectClaimableVouchers(
  channels: Channel[],
  opts?: SelectClaimableVouchersOptions,
): BatchSettlementVoucherClaim[] {
  const now = opts?.now ?? Date.now();
  const claims: BatchSettlementVoucherClaim[] = [];

  for (const c of channels) {
    if (BigInt(c.chargedCumulativeAmount) <= BigInt(c.totalClaimed)) {
      continue;
    }
    if (opts?.idleSecs !== undefined) {
      const idleMs = now - c.lastRequestTimestamp;
      if (idleMs < opts.idleSecs * 1000) {
        continue;
      }
    }
    claims.push({
      voucher: {
        channel: c.channelConfig,
        maxClaimableAmount: c.signedMaxClaimable,
      },
      signature: c.signature as `0x${string}`,
      totalClaimed: c.chargedCumulativeAmount,
    });
  }

  return claims;
}

/**
 * Updates session records after a successful claim so claim selection no longer
 * returns already-claimed vouchers.
 *
 * @param storage - Durable channel store.
 * @param claims - Voucher claims included in the submitted settlement transaction.
 * @param network - CAIP-2 network used to recompute channel ids.
 */
export async function applyClaimedTotals<T extends Channel = Channel>(
  storage: ChannelStorage<T>,
  claims: BatchSettlementVoucherClaim[],
  network: Network,
): Promise<void> {
  for (const claim of claims) {
    const channelId = computeChannelId(claim.voucher.channel, network);
    const channel = await storage.get(channelId);
    if (!channel) {
      continue;
    }
    const claimedAmount = BigInt(claim.totalClaimed);
    if (claimedAmount <= BigInt(channel.totalClaimed)) {
      continue;
    }
    await storage.updateChannel(channelId, current => {
      if (!current || claimedAmount <= BigInt(current.totalClaimed)) {
        return current;
      }
      return {
        ...current,
        totalClaimed: claimedAmount.toString(),
      };
    });
  }
}

File: go/mechanisms/evm/batch-settlement/storage/claims.go
package storage

import (
	"math/big"
	"time"

	batchsettlement "github.com/x402-foundation/x402/go/v2/mechanisms/evm/batch-settlement"
)

// SelectClaimableOptions is the optional wall-clock and idle filter for
// SelectClaimableVouchers.
type SelectClaimableOptions struct {
	Now      int64
	IdleSecs *int
}

// SelectClaimableVouchers collects vouchers that are eligible for onchain claiming.
//
// A voucher is claimable when its chargedCumulativeAmount exceeds what has
// already been claimed onchain. An optional idle filter skips sessions that
// received a request within the last IdleSecs seconds. Input order is preserved.
func SelectClaimableVouchers(channels []*Channel, opts *SelectClaimableOptions) []batchsettlement.BatchSettlementVoucherClaim {
	now := time.Now().UnixMilli()
	if opts != nil && opts.Now != 0 {
		now = opts.Now
	}
	claims := make([]batchsettlement.BatchSettlementVoucherClaim, 0)
	for _, c := range channels {
		if c == nil {
			continue
		}
		if uint256Cmp(c.ChargedCumulativeAmount, c.TotalClaimed) <= 0 {
			continue
		}
		if opts != nil && opts.IdleSecs != nil {
			idleMs := now - c.LastRequestTimestamp
			if idleMs < int64(*opts.IdleSecs)*1000 {
				continue
			}
		}
		claim := batchsettlement.BatchSettlementVoucherClaim{
			Signature:    c.Signature,
			TotalClaimed: c.ChargedCumulativeAmount,
		}
		claim.Voucher.Channel = c.ChannelConfig
		claim.Voucher.MaxClaimableAmount = c.SignedMaxClaimable
		claims = append(claims, claim)
	}
	return claims
}

// ApplyClaimedTotals updates session records after a successful claim so claim
// selection no longer returns already-claimed vouchers.
//
// The callback is authoritative: it re-checks row presence and the claimed
// watermark, so a claim that did not advance totalClaimed costs one no-op
// update rather than a pre-read on every claim.
func ApplyClaimedTotals[T ChannelRecord[T]](store ChannelStorage[T], claims []batchsettlement.BatchSettlementVoucherClaim, network string) error {
	for _, claim := range claims {
		channelId, err := batchsettlement.ComputeChannelId(claim.Voucher.Channel, network)
		if err != nil {
			return err
		}
		claimedAmount, ok := new(big.Int).SetString(claim.TotalClaimed, 10)
		if !ok {
			claimedAmount = new(big.Int)
		}
		if _, err := store.UpdateChannel(channelId, func(current T) T {
			if isZeroRecord(current) {
				return current
			}
			currentClaimed, ok := new(big.Int).SetString(current.Base().TotalClaimed, 10)
			if !ok {
				currentClaimed = new(big.Int)
			}
			if claimedAmount.Cmp(currentClaimed) <= 0 {
				return current
			}
			next := current.Clone()
			next.Base().TotalClaimed = claimedAmount.String()
			return next
		}); err != nil {
			return err
		}
	}
	return nil
}
